[Feature] Data sources associated objects tab

RyanL1997 opened the pull request:
### Description
Add Data sources associated objects tab

### Issues Resolved
[List any issues this PR will resolve]

### Demo
Without data:
![Screenshot 2024-02-26 at 4 18 47 PM](https://github.com/opensearch-project/dashboards-observability/assets/109499885/fe5937d5-0e98-43eb-bb04-5cee1eb7a430)

With data:
![Screenshot 2024-02-26 at 4 19 40 PM](https://github.com/opensearch-project/dashboards-observability/assets/109499885/1dd1e528-6bc3-407f-8e39-cbd263b90af2)


### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass, including unit test, integration test and doctest
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
  - [ ] New functionality has user manual doc added
- [x] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

RyanL1997 commented on opensearch_dashboards.json:
```diff
@@ -1,7 +1,7 @@
 {
   "id": "observabilityDashboards",
-  "version": "3.0.0.0",
```
this will be removed

RyanL1997 commented on public/components/datasources/components/manage/data_connection.tsx:
```diff
@@ -165,6 +166,35 @@ export const DataConnection = (props: any) => {
+        <AssociatedObjectsTab
+          associatedObjects={[
+            {
+              id: '1',
```
@ps48 do you want me to remove this sample data? Or how are we gonna test this?

ps48 replied:
Let's keep the sample data here.

sejli commented on public/components/datasources/components/manage/associated_objects_tab.tsx:
```diff
@@ -0,0 +1,176 @@
+
+  const pagination = {
+    initialPageSize: 10,
+    pageSizeOptions: [5, 10, 20],
```
Nitpick: We can probably increase this to `10, 25, 50` in some other [OSD components](https://github.com/search?q=org%3Aopensearch-project%20pageSizeOptions&type=code)

sejli commented on public/components/datasources/components/manage/associated_objects_tab.tsx:
```diff
@@ -0,0 +1,176 @@
+        <EuiInMemoryTable
+          items={associatedObjects}
+          columns={columns}
+          search={search}
```
May need to move this out for an external search bar component since the `EuiInMemoryTable` doesn't have support for search bar **and** filters, but that can be done in a future PR.

RyanL1997 replied:
lets do this as a follow up

sejli commented on public/components/datasources/components/manage/associated_objects_tab.tsx:
```diff
@@ -0,0 +1,176 @@
+} from '@elastic/eui';
+import { AccelerationsRecommendationCallout } from './accelerations_recommendation_callout';
+
+interface AssociatedObject {
+  id: string;
+  name: string;
+  database: string;
+  type: string;
+  createdByIntegration: string;
+  accelerations: string;
+}
```
Could be exported out of [`dashboards-observability/common/types/data_connections.ts`](https://github.com/opensearch-project/dashboards-observability/blob/main/common/types/data_connections.ts), since other files could use the same object type.

RyanL1997 replied:
done

sejli commented:
In order to fix test and build workflows, `yarn test -u` should be run to update the snapshots. That should fix those workflows.

ps48 reviewed (commented):
@RyanL1997 For the below lint error let's add a `console.error` to log the error whenever it occurs. 
```
136:15  error    'err' is defined but never used. Allowed unused args must match /^_/u   
```
Thanks! for quick the updates on the PR

RyanL1997 commented:
@ps48 @sejli , I have addressed the comments above and also fixed that lint ( that was irrelevant from this change btw. ). And I'm updating the screenshots to the description.

joshuali925 commented on common/types/data_connections.ts:
```diff
@@ -13,6 +13,15 @@ export interface PermissionsConfigurationProps {
+  database: string;
+  type: string;
+  createdByIntegration: string;
+  accelerations: string;
```
would any of these fields be a smaller type than `string`, or all of them can be arbitrary strings?

joshuali925 commented on public/components/datasources/components/manage/associated_objects_tab.tsx:
```diff
@@ -0,0 +1,197 @@
+        <AccelerationsRecommendationCallout />
+        <EuiSpacer />
+        {associatedObjects.length > 0 ? (
+          <EuiInMemoryTable
```
would there be a case that `associatedObjects` gets large and doesn't fit in memory?

RyanL1997 replied:
Good point, I think for the memory fitting issue, we may need to do something like virtualization with react-window to limit the size of the list. However, since this this is a skeleton for now, and lets just leave like this. I will test it when we connect the actual data content, so that we can decide which approaches we need for addressing the memory issue. I will not resolve this comment btw until we have a concrete solution on this.

Swiddis commented on common/types/data_connections.ts:
```diff
@@ -13,6 +13,15 @@ export interface PermissionsConfigurationProps {
+  name: string;
+  database: string;
+  type: string;
+  createdByIntegration: string;
```
I don't know if `createdByIntegration` is a field that `AssociatedObject` would contain, I'd expect this to be queried from the integrations that store all the contained object information. Do you see us modifying Integrations to add fields like this to arbitrary objects? If so, it might be worth adding this requirement in #1442

RyanL1997 replied:
Transferring some of the discussion over here: for now, since we should able to have this field with the source of integration, lets have this field for now for this skeleton PR. We can sync up on this again when we actually do the connection of actual data content.
